File: Day_29/Q116.java
import java.util.*;
class Q116{
    static Scanner sc = new Scanner(System.in);

    static class Product{
        int id;
        String name;
        int quantity;
        double price;
    }

    //Add Product
    static void add_product(ArrayList<Product> products){
        Product p = new Product();
        System.out.print("Enter product id: ");
        p.id = sc.nextInt();
        sc.nextLine();
        System.out.print("Enter product name: ");
        p.name = sc.nextLine();
        System.out.print("Enter quantity: ");
        p.quantity = sc.nextInt();
        System.out.print("Enter price: ");
        p.price = sc.nextDouble();
        products.add(p);
    }

    //Display Products
    static void display_products(ArrayList<Product> products){
        if(products.size() == 0){
            System.out.println("No products available");
            return;
        }
        for(Product p : products){
            System.out.println("ID: " + p.id);
            System.out.println("Name: " + p.name);
            System.out.println("Quantity: " + p.quantity);
            System.out.println("Price: " + p.price);
            System.out.println();
        }
    }

    //Search Product
    static void search_product(ArrayList<Product> products){
        int id;
        System.out.print("Enter product id: ");
        id = sc.nextInt();
        for(Product p : products){
            if(p.id == id){
                System.out.println("Product Found");
                System.out.println("Name: " + p.name);
                System.out.println("Quantity: " + p.quantity);
                System.out.println("Price: " + p.price);
                return;
            }
        }
        System.out.println("Product not found");
    }

    //Update Quantity
    static void update_quantity(ArrayList<Product> products){
        int id,qty;
        System.out.print("Enter product id and new quantity: ");
        id = sc.nextInt();
        qty = sc.nextInt();
        for(int i = 0;i < products.size();i++){
            if(products.get(i).id == id){
                products.get(i).quantity = qty;
                System.out.println("Quantity updated");
                return;
            }
        }
        System.out.println("Product not found");
    }

    //Menu
    static void menu(){
        System.out.println("\n1. Add Product");
        System.out.println("2. Display Products");
        System.out.println("3. Search Product");
        System.out.println("4. Update Quantity");
        System.out.println("5. Exit");
        System.out.print("Enter your choice: ");
    }

    public static void main(String[] args){
        ArrayList<Product> products = new ArrayList<Product>();
        int choice;
        do{
            menu();
            choice = sc.nextInt();
            switch(choice){
                case 1: add_product(products); break;
                case 2: display_products(products); break;
                case 3: search_product(products); break;
                case 4: update_quantity(products); break;
                case 5: System.out.println("Program ended"); break;
                default: System.out.println("Invalid choice");
            }
        }while(choice != 5);
        sc.close();
    }
}

File: Day_29/Q116.cpp
#include <bits/stdc++.h>
using namespace std;

struct Product{
    int id;
    string name;
    int quantity;
    double price;
};

//Add Product
void add_product(vector<Product> &products){
    Product p;
    cout << "Enter product id: ";
    cin >> p.id;
    cin.ignore();
    cout << "Enter product name: ";
    getline(cin,p.name);
    cout << "Enter quantity: ";
    cin >> p.quantity;
    cout << "Enter price: ";
    cin >> p.price;
    products.push_back(p);
}

//Display Products
void display_products(vector<Product> &products){
    if(products.size() == 0){
        cout << "No products available" << endl;
        return;
    }
    for(Product p : products){
        cout << "ID: " << p.id << endl;
        cout << "Name: " << p.name << endl;
        cout << "Quantity: " << p.quantity << endl;
        cout << "Price: " << p.price << endl << endl;
    }
}

//Search Product
void search_product(vector<Product> &products){
    int id;
    cout << "Enter product id: ";
    cin >> id;
    for(Product p : products){
        if(p.id == id){
            cout << "Product Found" << endl;
            cout << "Name: " << p.name << endl;
            cout << "Quantity: " << p.quantity << endl;
            cout << "Price: " << p.price << endl;
            return;
        }
    }
    cout << "Product not found" << endl;
}

//Update Quantity
void update_quantity(vector<Product> &products){
    int id,qty;
    cout << "Enter product id and new quantity: ";
    cin >> id >> qty;
    for(int i = 0;i < products.size();i++){
        if(products[i].id == id){
            products[i].quantity = qty;
            cout << "Quantity updated" << endl;
            return;
        }
    }
    cout << "Product not found" << endl;
}

//Menu
void menu(){
    cout << "\n1. Add Product" << endl;
    cout << "2. Display Products" << endl;
    cout << "3. Search Product" << endl;
    cout << "4. Update Quantity" << endl;
    cout << "5. Exit" << endl;
    cout << "Enter your choice: ";
}

int main(){
    vector<Product> products;
    int choice;
    do{
        menu();
        cin >> choice;
        switch(choice){
            case 1: add_product(products); break;
            case 2: display_products(products); break;
            case 3: search_product(products); break;
            case 4: update_quantity(products); break;
            case 5: cout << "Program ended" << endl; break;
            default: cout << "Invalid choice" << endl;
        }
    }while(choice != 5);
    return 0;
}
